### PR TITLE
implement api function for get_vote_info

### PR DESCRIPTION
**Pull request by energyGiver**
### Summary
* implement a runtime api function to get the VoteInfo
* chagne the VoteInfo type from StorageMap to CountedStorageMap

### Describe your changes
* implement a get_vote_info function to PoT pallet and node_template runtime.
* design VoteInfo and get_vote_info function to bound and efficiently compute the length.
* implement a test function when the number of selected validators has exceeded the maximum allowed limit.

### Related Issue
* #8

### Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] Well documented code
- [ ] `cargo clippy`
- [x] `cargo-nightly fmt`

### CODE REVIEW

**Comment by cocoyoon** [comment]
@energyGiver Sirius 도 review 에 넣어주세요
Assignee 본인으로 설정해주시고요

File: frame/proof-of-transaction/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+sp_api::decl_runtime_apis! {
+	#[api_version(3)]
+	pub trait ProofOfTransactionAPI<AccountId> where
+	AccountId: codec::Codec,
+	{
+		fn get_vote_info(account: AccountId) -> Vec<(AccountId, u64)>;
+	}
+}
```
Comment (cocoyoon) [inline frame/proof-of-transaction/src/lib.rs:186]: Parity substrate 디렉토리 참고하셔서 runtime_api 모듈에 넣어주세요
- `proof-of-transaction/runtime-api/lib.rs`

File: frame/proof-of-transaction/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		VoteCollected { candidate: T::AccountId, weight: VoteWeight },
+		NoVote,
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn get_vote_info(account: T::AccountId) -> Vec<(T::AccountId, VoteWeight)> {
+			VoteInfo::<T>::iter().collect::<Vec<(T::AccountId, VoteWeight)>>()
+		}
+	}
+}
```
Comment (cocoyoon) [inline frame/proof-of-transaction/src/lib.rs:69]: 1. `mod pallet` 밖에서 정의해주세요!
2. `BoundedVec` 까지 해서 merge 하도록 하죠

**Comment by cocoyoon** [comment]
@sweatpotato13 현재 `master` 에 바로 merge 하게 돼있는데 `develop` 으로 변경하려면 새로 pr 다시 해야되겟죠?

**Comment by sweatpotato13** [comment]
> @sweatpotato13 현재 `master` 에 바로 merge 하게 돼있는데 `develop` 으로 변경하려면 새로 pr 다시 해야되겟죠?

Edit으로 base branch 변경할 수 있어요

File: frame/proof-of-transaction/Cargo.toml [hunk 3]
```diff
@@ -8,11 +8,18 @@ homepage = "https://bc-labs.net"
...
 
+[package.metadata.docs.rs]
+targets = ["x86_64-unknown-linux-gnu"]
+
 [dependencies]
 
 # Other dependencies
-codec = { package = "parity-scale-codec", version = "3.2.2", default-features = false, features = ["derive"] }
-scale-info = { version = "2.1.1", default-features = false, features = ["derive"] }
+codec = { package = "parity-scale-codec", version = "3.2.2", default-features = false, features = [
```
Comment (sweatpotato13) [inline frame/proof-of-transaction/Cargo.toml:17]: Cargo.toml은 `cargo +nightly fmt` 수행하면서 바뀐건가요?

File: frame/proof-of-transaction/src/lib.rs (hunk 1 above)
Comment (energyGiver) [reply frame/proof-of-transaction/src/lib.rs:186]: > 

changed it!

File: frame/proof-of-transaction/Cargo.toml (hunk 3 above)
Comment (energyGiver) [reply frame/proof-of-transaction/Cargo.toml:17]: Seems to be a problem with my VS code auto indent. I'll fix it.

Comment (sweatpotato13) [reply frame/proof-of-transaction/Cargo.toml:17]: * frame/proof-of-transaction/Cargo.toml
* frame/transaction-payment/rpc/runtime-api/Cargo.toml

Comment (energyGiver) [reply frame/proof-of-transaction/Cargo.toml:17]: reflected your comment!

File: frame/proof-of-transaction/src/lib.rs (hunk 1 above)
Comment (energyGiver) [reply frame/proof-of-transaction/src/lib.rs]: Addressed your comment!

File: Cargo.lock
[the remark this answers is not in the document]
Comment (sweatpotato13) [reply Cargo.lock:1]: cargo.lock 지우고 다시 빌드해서 생긴 cargo.lock 파일로 대체하면 될 것 같아요

File: frame/proof-of-transaction/src/lib.rs [hunk 4]
```diff
@@ -6,19 +6,26 @@ mod mock;
...
 	transaction_validity::{TransactionValidity, TransactionValidityError, ValidTransaction},
 };
 
-/// Type of weight that refers to 'ref_time' in Weight struct
+// use sp_runtime::traits::{Bounded, Hash, StaticLookup};
+use sp_std::{convert::TryInto, prelude::*};
+
+use frame_support::BoundedVec;
+
+use sp_std::vec::Vec;
```
[the remark this answers is not in the document]
Comment (energyGiver) [reply frame/proof-of-transaction/src/lib.rs:27]: done

File: bin/node-template/runtime/src/lib.rs [hunk 5]
```diff
@@ -475,6 +477,13 @@ impl_runtime_apis! {
 		}
 	}
 
+	impl proof_of_transaction_runtime_api::ProofOfTransactionAPI<Block, AccountId, <Runtime as pallet_pot::Config>::MaxValidators> for Runtime {
+		fn get_vote_info() -> BoundedVec<(AccountId, u64), MaxValidators> {
```
[the remark this answers is not in the document]
Comment (energyGiver) [reply bin/node-template/runtime/src/lib.rs:481]: done

File: frame/proof-of-transaction/src/lib.rs [hunk 6]
```diff
@@ -41,6 +44,9 @@ pub mod pallet {
 	pub trait Config: frame_system::Config {
 		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
 
+		#[pallet::constant]
+		type MaxVotedValidators: Get<u32>;
+
```
Comment (cocoyoon) [inline frame/proof-of-transaction/src/lib.rs:49]: Doc 추가해주세욤

File: frame/proof-of-transaction/runtime-api/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+use frame_support::{traits::Get, BoundedVec};
+
+sp_api::decl_runtime_apis! {
+	#[api_version(3)]
+	pub trait ProofOfTransactionAPI<AccountId, MaxVotedValidators: Get<u32>> where
+		AccountId: codec::Codec,
+	{
+		fn get_vote_info() -> BoundedVec<(AccountId, u64), MaxVotedValidators>;
+	}
```
Comment (cocoyoon) [inline frame/proof-of-transaction/runtime-api/src/lib.rs:30]: Doc 추가하셨나요??

File: frame/proof-of-transaction/src/tests.rs [hunk 8]
```diff
@@ -79,3 +79,32 @@ fn signed_extension_works() {
 		System::assert_has_event(TestEvent::Pot(Event::NoVote));
 	})
 }
+
+#[test]
+// #[should_panic]
+fn collect_vote_for_should_fail() {
```
Comment (cocoyoon) [inline frame/proof-of-transaction/src/tests.rs:85]: 어디가 패닉 나는 거에요?

File: frame/proof-of-transaction/src/lib.rs [hunk 9]
```diff
@@ -51,6 +57,10 @@ pub mod pallet {
 	pub type VoteInfo<T: Config> =
 		StorageMap<_, Twox64Concat, T::AccountId, VoteWeight, OptionQuery>;
```
[the remark this answers is not in the document]
Comment (energyGiver) [reply frame/proof-of-transaction/src/lib.rs:59]: right, deleted it

File: frame/proof-of-transaction/src/lib.rs (hunk 6 above)
Comment (energyGiver) [reply frame/proof-of-transaction/src/lib.rs:49]: would you explain about "Doc"?

File: frame/proof-of-transaction/runtime-api/src/lib.rs (hunk 7 above)
Comment (energyGiver) [reply frame/proof-of-transaction/runtime-api/src/lib.rs:30]: done

File: frame/proof-of-transaction/src/lib.rs (hunk 6 above)
Comment (energyGiver) [reply frame/proof-of-transaction/src/lib.rs:49]: done

**Review by sweatpotato13 — APPROVED** [review]
👍

**Pull request merged**